### ISSUE TITLE
Segmentation Fault in libcaffe2.so (called during torch.tensor())

### ISSUE THREAD

**Issue by ecobost** [pytorch/pytorch#23213]
## 🐛 Bug

Segmentation fault when calling torch.tensor for specific numpy arrays (or torch.clone or any call that copies the storage of the numpy array).

It may be related to the array not being memory aligned or being loaded using np.frombuffer (thus being readonly with its data being the bytes of the string); however, it does not fail for other arrays stored in the exact same way.

## To Reproduce
When I saved and reloaded the array (using pickle or np.tostring or np.save), the error will stop happening (maybe because the reloaded array will be memory aligned) so I cannot send a sample to reproduce this, but here's some info about the offending array:
```python3
In [8]: b                                                                                                                                                                                                          
Out[8]: 
array([ 0.24723893, -0.02146482,  0.00140578, ..., -0.05349024,
        0.18419194, -0.50951517], dtype=float32)

In [9]: np.isnan(b).any()                                                                                                                                                                                          
Out[9]: False

In [10]: np.isinf(b).any()                                                                                                                                                                                         
Out[10]: False

In [4]: b.flags                                                                                                                                                                                                    
Out[4]: 
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : False
  WRITEABLE : False
  ALIGNED : False
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False

In [6]: b.strides                                                                                                                                                                                                  
Out[6]: (4,)

In [19]: b.data                                                                                                                                                                                                    
Out[19]: <memory at 0x7fc6ffd2ec48>

In [18]: b.base                                                                                                                                                                                                    
Out[18]: 
array([ 0.24723893, -0.02146482,  0.00140578, ..., -0.05349024,
        0.18419194, -0.50951517], dtype=float32)

In [20]: b.base.data                                                                                                                                                                                               
Out[20]: <memory at 0x7fc703c14a08>

In [21]: b.base.base # this is the byte representation of the dictionary from where I loaded b
Out[21]:
b'mYm\x00S\x01\x00\x00\x00\x00......
```

Here's a summary of where the error happens and when it does not:
```python3
b = np.zeros(8127, dtype=np.float32) # fake array

# Works
torch.tensor(b[:-1]) 
torch.tensor(b[1:])
torch.from_numpy(b) # no copying, new tensor and b share the same storage
torch.as_tensor(b) # no copying, new tensor and b share the same storage
torch.as_tensor(b, device='cuda')
torch.tensor(b.copy()) # creating a copy solves it but why?

# Any of this lines causes a segmentation fault and kills the kernel
torch.tensor(b)
torch.tensor(b, device='cpu', dtype=torch.float32, requires_grad=False)
torch.tensor(b[:len(b)])
torch.from_numpy(b).clone()
torch.zeros(8127).copy_(torch.from_numpy(b)) # a similar line is executed inside load_state_dict
torch.tensor(b.base)
b.setflags(write=True); torch.tensor(b) # with numpy 1.15.0
```

Here's the stack trace from gdb (test.py just loads the array from our database and calls torch.tensor on it):
```gdb
(gdb) run test.py
Starting program: /usr/bin/python3 test.py
warning: Error disabling address space randomization: Operation not permitted
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7f8d9d757700 (LWP 4157)]
[New Thread 0x7f8d9cf56700 (LWP 4158)]
[New Thread 0x7f8d98755700 (LWP 4159)]
[New Thread 0x7f8d95f54700 (LWP 4160)]
[New Thread 0x7f8d93753700 (LWP 4161)]
[New Thread 0x7f8d90f52700 (LWP 4162)]
[New Thread 0x7f8d8e751700 (LWP 4163)]
[New Thread 0x7f8d8bf50700 (LWP 4164)]
[New Thread 0x7f8d8b74f700 (LWP 4165)]
[New Thread 0x7f8d88f4e700 (LWP 4166)]
[New Thread 0x7f8d8474d700 (LWP 4167)]
[Thread 0x7f8d8b74f700 (LWP 4165) exited]
[Thread 0x7f8d88f4e700 (LWP 4166) exited]
[Thread 0x7f8d8bf50700 (LWP 4164) exited]
[Thread 0x7f8d8e751700 (LWP 4163) exited]
[Thread 0x7f8d90f52700 (LWP 4162) exited]
[Thread 0x7f8d8474d700 (LWP 4167) exited]
[Thread 0x7f8d93753700 (LWP 4161) exited]
[Thread 0x7f8d95f54700 (LWP 4160) exited]
[Thread 0x7f8d98755700 (LWP 4159) exited]
[Thread 0x7f8d9cf56700 (LWP 4158) exited]
[Thread 0x7f8d9d757700 (LWP 4157) exited]
Connecting ecobost@at-database.ad.bcm.edu:3306

Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
0x00007f8d26a76758 in std::_Function_handler<void (int, char**, long const*, long), void at::native::(anonymous namespace)::unary_kernel_vec<at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#4}::operator()() const::{lambda(float)#1}, at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#2}::operator()() const::{lambda(at::vec256::(anonymous namespace)::Vec256<float>)#2}>(at::TensorIterator&, at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#4}::operator()() const::{lambda(float)#1}, at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#2}::operator()() const::{lambda(at::vec256::(anonymous namespace)::Vec256<float>)#2})::{lambda(int, char**, long const*, long)#1}>::_M_invoke(std::_Any_data const&, int, char**, long const*, long) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
(gdb) where
#0  0x00007f8d26a76758 in std::_Function_handler<void (int, char**, long const*, long), void at::native::(anonymous namespace)::unary_kernel_vec<at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#4}::operator()() const::{lambda(float)#1}, at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#2}::operator()() const::{lambda(at::vec256::(anonymous namespace)::Vec256<float>)#2}>(at::TensorIterator&, at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#4}::operator()() const::{lambda(float)#1}, at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&)::{lambda()#2}::operator()() const::{lambda()#2}::operator()() const::{lambda(at::vec256::(anonymous namespace)::Vec256<float>)#2})::{lambda(int, char**, long const*, long)#1}>::_M_invoke(std::_Any_data const&, int, char**, long const*, long) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#1  0x00007f8d25fcf906 in std::_Function_handler<void (int, char**, long const*, long, long), at::loop_wrapper(std::function<void (int, char**, long const*, long)> const&)::{lambda(int, char**, long const*, long, long)#1}>::_M_invoke(std::_Any_data const&, int, char**, long const*, long, long) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#2  0x00007f8d25fd2558 in at::TensorIterator::serial_for_each(std::function<void (int, char**, long const*, long, long)> const&, at::Range) const ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#3  0x00007f8d25fd2790 in at::TensorIterator::for_each(std::function<void (int, char**, long const*, long, long)> const&) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#4  0x00007f8d25fd289f in at::TensorIterator::for_each(std::function<void (int, char**, long const*, long)> const&) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#5  0x00007f8d26a7cb97 in at::native::(anonymous namespace)::copy_kernel_same_type_impl(at::Tensor&, at::Tensor const&) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#6  0x00007f8d25e7ffbe in at::native::_s_copy__cpu(at::Tensor&, at::Tensor const&, bool) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#7  0x00007f8d26141b9f in at::CPUType::s_copy_(at::Tensor&, at::Tensor const&, bool) const () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#8  0x00007f8d25e7f33d in at::native::copy_(at::Tensor&, at::Tensor const&, bool) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#9  0x00007f8d243f9dc9 in torch::autograd::VariableType::copy_(at::Tensor&, at::Tensor const&, bool) const () from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1
#10 0x00007f8d25fc07c6 in at::native::to(at::Tensor const&, c10::Device, c10::ScalarType, bool, bool) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#11 0x00007f8d2624087b in at::TypeDefault::to(at::Tensor const&, c10::Device, c10::ScalarType, bool, bool) const () from /usr/local/lib/python3.6/dist-packages/torch/lib/libcaffe2.so
#12 0x00007f8d241d000d in torch::autograd::VariableType::to(at::Tensor const&, c10::Device, c10::ScalarType, bool, bool) const () from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch.so.1
#13 0x00007f8d6dd297ce in torch::utils::(anonymous namespace)::internal_new_from_data(at::Type const&, c10::ScalarType, c10::optional<c10::Device>, _object*, bool, bool, bool, bool) ()
   from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_python.so
#14 0x00007f8d6dd2a98d in torch::utils::tensor_ctor(at::Type const&, c10::ScalarType, _object*, _object*) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_python.so
#15 0x00007f8d6d9f1b13 in torch::autograd::THPVariable_tensor(_object*, _object*, _object*) () from /usr/local/lib/python3.6/dist-packages/torch/lib/libtorch_python.so
#16 0x00000000004f8925 in _PyCFunction_FastCallDict (kwargs=<optimized out>, nargs=<optimized out>, args=0x17be878, func_obj=<built-in method tensor of type object at remote 0x7f8d6e32eb80>)
    at ../Objects/methodobject.c:231
#17 _PyCFunction_FastCallKeywords (kwnames=<optimized out>, nargs=<optimized out>, stack=<optimized out>, func=<optimized out>) at ../Objects/methodobject.c:294
#18 call_function () at ../Python/ceval.c:4851
#19 0x00000000004f98c7 in _PyEval_EvalFrameDefault () at ../Python/ceval.c:3335
#20 0x00000000004f6128 in PyEval_EvalFrameEx (throwflag=0, f=Frame 0x17be6f8, for file test.py, line 16, in <module> ()) at ../Python/ceval.c:4166
#21 _PyEval_EvalCodeWithName.lto_priv.1581 () at ../Python/ceval.c:4166
#22 0x00000000004f9023 in PyEval_EvalCodeEx (closure=0x0, kwdefs=0x0, defcount=0, defs=0x0, kwcount=0, kws=0x0, argcount=0, args=0x0, locals=<optimized out>, globals=<optimized out>, _co=<optimized out>)
    at ../Python/ceval.c:4187
#23 PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at ../Python/ceval.c:731
#24 0x00000000006415b2 in run_mod () at ../Python/pythonrun.c:1025
#25 0x000000000064166a in PyRun_FileExFlags () at ../Python/pythonrun.c:978
#26 0x0000000000643730 in PyRun_SimpleFileExFlags () at ../Python/pythonrun.c:419
#27 0x000000000062b26e in run_file (p_cf=0x7fffe1582bbc, filename=<optimized out>, fp=<optimized out>) at ../Modules/main.c:340
#28 Py_Main () at ../Modules/main.c:810
#29 0x00000000004b4cb0 in main (argc=2, argv=0x7fffe1582db8) at ../Programs/python.c:69
(gdb) 
```
The other lines also error in the same copy_kernel_same_type_impl() inside libcaffe2.so.

## Environment
```
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: TITAN V
GPU 1: GeForce GT 710

Nvidia driver version: 430.26
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.1.0
[pip3] torchvision==0.3.0
[conda] Could not collect
```
- Any other relevant information:
I am running inside a Docker image (atlab/pytorch in dockerhub). 
Error also occurs with numpy 1.15.0 and with the latest nightly build of pytorch ('1.2.0.dev20190722'). Does NOT occur in pytorch 0.4.0 (CUDA 9.1)

## Additional context
There are a couple other [Segmentation fault Issues](https://github.com/pytorch/pytorch/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+Segmentation+fault+libcaffe2) that may be related but I'm not sure

**Comment by ecobost** [comment]
Could it be related to https://github.com/pytorch/pytorch/pull/15845

**Comment by gchanan** [comment]
I wasn't able to reproduce, I get:
```
Traceback (most recent call last):
  File "fooyaz.py", line 19, in <module>
    torch.zeros(8127).copy_(torch.from_numpy(b)) # a similar line is executed inside load_state_dict
RuntimeError: The size of tensor a (8127) must match the size of tensor b (3) at non-singleton dimension 0
```

@ecobost can you try against master?

**Comment by ecobost** [comment]
@gchanan Built from master (version ''1.2.0a0+d8220b0'). Error is still there. Code above won't reproduce it because it only happens with (some of) the arrays that we saved and reloaded from our database, code in there does cause segmentation faults for only for these arrays. I can't find a way to save one of the offending arrays in a way that it will still cause the segmentation fault after reloading, otherwise I would certainly share it.

**Comment by mruberry** [comment]
@ecoboost I'm doing a survey of numpy-related issues. Do you still experience this issue? Did you find a workaround? 

Also, can you describe how the tensors that cause segmentation faults were constructed?

**Comment by ecobost** [comment]
Just retested it, could not replicate. The workaround was to copy the numpy array before sending it to `as_tensor()`. 
The arrays were loaded from our database. We use a library (https://github.com/datajoint/datajoint-python) to interface with the database and I ignore how the serialization and deserialization of the numpy arrays happens. I guess something changed there; version of pytorch and numpy are still the same but I did not record the version of this library so I don't know if it has been updated.
Feel free to close this issue
